core/src/main/scala/com/pingcap/tispark/TiBatchWrite.scala

```diff
@@ -98,10 +102,30 @@ class TiBatchWrite(@transient val df: DataFrame,
     if (tiTableInfo == null) {
       throw new NoSuchTableException(tiTableRef.databaseName, tiTableRef.tableName)
     }
+    dfColSize = df.columns.length
+    tableColSize = tiTableInfo.getColumns.size()
 
     // check check check
     checkUnsupported()
     checkColumnNumbers()
+
+    // initialize dfTiColumnInfo
+    dfTiColumnInfo = new Array[TiColumnInfo](dfColSize)
```

Comment:
this should move to sparkRow2TiRow. Otherwise, we do not have a way to allocate id.